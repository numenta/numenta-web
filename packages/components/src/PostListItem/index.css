/**
 * Numenta Web Platform and Sites source code
 * MIT License (see LICENSE.txt)
 * Copyright © 2005—2017 Numenta <http://numenta.com>
 */

.item {
  /* */
}

.date {
  composes: fw5 from 'tachyons';
  composes: lh-copy from 'tachyons';
  composes: mb2 from 'tachyons';
}

.title {
  composes: db from 'tachyons';
  composes: dim from 'tachyons';
  composes: link from 'tachyons';
  composes: f4 from 'tachyons';
  composes: lh-title from 'tachyons';
  composes: mb1 from 'tachyons';
}

.event {
  composes: bg-near-white from 'tachyons';
  composes: br2 from 'tachyons';
  composes: mb2 from 'tachyons';
  composes: mr3-ns from 'tachyons';
  composes: mt3 from 'tachyons';
  composes: pa2 from 'tachyons';
}
.event table {
  margin: 0;
  padding: 0;
}

.author {
  composes: lh-copy from 'tachyons';
}

.image {
  composes: center from 'tachyons';
  composes: w-80 from 'tachyons';
  composes: w-auto-m from 'tachyons';
}
