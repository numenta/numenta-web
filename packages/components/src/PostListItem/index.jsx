// Numenta Web Platform and Sites source code
// MIT License (see LICENSE.txt)
// Copyright © 2005—2017 Numenta <http://numenta.com>

import {getEventTimeDisplay} from 'numenta-web-shared-utils/lib/shared'
import moment from 'moment'
import React from 'react'

import Avatar from '../Avatar'
import Paragraph from '../Paragraph'
import Spacer from '../Spacer'
import Strong from '../Strong'
import Subtle from '../Subtle'
import Table from '../Table'
import TableBody from '../TableBody'
import TableCell from '../TableCell'
import TableRow from '../TableRow'
import TextLink from '../TextLink'
import Time from '../Time'

import styles from './index.css'


/**
 * Post Item Row - React view component for Post List.
 */
const PostListItem = ({post}, {config}) => {
  const {data, file, path} = post
  const datetime = moment(data.date, config.moments.post)
  const occur = datetime.format(config.moments.human)
  const target = (data.type === 'link') ? data.link : path
  const key = file.dir.split('/')[0]
  let brief = (<Paragraph>{data.brief}...</Paragraph>)

  // modifications for: events
  if (key === 'events') {
    const {when, where} = data.event
    const {desc, city, state, country} = where
    let location = city

    if (state) {
      location = `${location}, ${state}`
    }
    if (country) {
      location = `${location} ${country}`
    }

    brief = (
      <div className={styles.event}>
        <Table direction="horizontal">
          <TableBody>
            <TableRow>
              <TableCell>
                <Strong>When</Strong>
              </TableCell>
              <TableCell>
                {getEventTimeDisplay(when)}
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <Strong>Where</Strong>
              </TableCell>
              <TableCell>
                <div>{desc}</div>
                <div>
                  {location}
                </div>
              </TableCell>
            </TableRow>
          </TableBody>
        </Table>
      </div>
    )
  }

  return (
    <div className={styles.item}>
      <div className={styles.date}>
        <Time moment={datetime}>{occur}</Time>
      </div>
      <TextLink to={target}>
        <span className={styles.title}>
          {data.title}
        </span>
      </TextLink>
      {brief}
      <div className={styles.author}>
        <Subtle>
          <Avatar name={data.author} />
          {data.author}
          <Spacer />
          {data.org}
        </Subtle>
      </div>
    </div>
  )
}

PostListItem.contextTypes = {
  config: React.PropTypes.object,
}

PostListItem.propTypes = {
  post: React.PropTypes.object.isRequired,
}

export default PostListItem
