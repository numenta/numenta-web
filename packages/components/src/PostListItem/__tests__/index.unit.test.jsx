import merge from 'lodash/merge'
import React from 'react'
import renderer from 'react-test-renderer'
import stubContext from 'react-stub-context'

import context from '__tests__/__mocks__/reactContextMock'
import post from '__tests__/__mocks__/postMock'

import PostListItem from '../../PostListItem'


const PostListItemStubbed = stubContext(PostListItem, context)

const eventMock = {
  when: {
    begin: '2016/08/20 13:00',
    end: '2016/08/20 14:00',
  },
  where: {
    desc: 'Big Ben',
    city: 'London',
    state: 'England',
    country: 'UK',
    web: 'http://bigben.co.uk',
  },
}
const linkMock = '/press/2015/08/20/numenta-releases-product/'


describe('PostListItem React component', () => {

  it('Renders Normal Post (Blog, Press Release, etc.) correctly', () => {
    const component = renderer.create(
      <PostListItemStubbed post={post}>
        Blog-like Content
      </PostListItemStubbed>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Renders Evented Post (Event) correctly', () => {
    const event = merge({}, post, eventMock)
    const component = renderer.create(
      <PostListItemStubbed post={event}>
        Event-like Content
      </PostListItemStubbed>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Renders Link stub (Press Link) correctly', () => {
    const stem = linkMock.split('/').pop()
    const link = merge({}, post, {
      data: {
        link: 'http://numenta.com',
        type: 'link',
      },
      file: {
        dir: linkMock,
        path: linkMock,
        stem,
      },
      key: stem,
      path: linkMock,
    })
    const component = renderer.create(
      <PostListItemStubbed post={link} />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

})
