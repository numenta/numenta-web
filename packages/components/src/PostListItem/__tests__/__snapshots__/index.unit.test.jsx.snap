exports[`PostListItem React component Renders Evented Post (Event) correctly 1`] = `
<div
  className="item">
  <div
    className="date">
    <time
      className="time"
      dateTime="2016-12-21T00:00:00-08:00">
      Wed, Dec 21, 2016
    </time>
  </div>
  <a
    className="textlink"
    href={undefined}
    onClick={[Function]}
    style={Object {}}
    target={undefined}>
    <span
      className="title">
      Calculus of the Nervous System
    </span>
  </a>
  <p
    className="paragraph">
    Quick quip or intro to the full text.
    ...
  </p>
  <div
    className="author">
    <span
      className="subtle">
      <span
        className="avatar">
        <img
          alt="Ada Lovelace"
          className="image border round"
          onClick={undefined}
          src="filename.ext"
          title="Ada Lovelace" />
      </span>
      Ada Lovelace
      <span
        className="spacer">
        •
      </span>
      Evangelist
    </span>
  </div>
</div>
`;

exports[`PostListItem React component Renders Link stub (Press Link) correctly 1`] = `
<div
  className="item">
  <div
    className="date">
    <time
      className="time"
      dateTime="2016-12-21T00:00:00-08:00">
      Wed, Dec 21, 2016
    </time>
  </div>
  <a
    className="textlink"
    href="http://numenta.com"
    onClick={[Function]}
    target={undefined}>
    <span
      className="title">
      Calculus of the Nervous System
    </span>
  </a>
  <p
    className="paragraph">
    Quick quip or intro to the full text.
    ...
  </p>
  <div
    className="author">
    <span
      className="subtle">
      <span
        className="avatar">
        <img
          alt="Ada Lovelace"
          className="image border round"
          onClick={undefined}
          src="filename.ext"
          title="Ada Lovelace" />
      </span>
      Ada Lovelace
      <span
        className="spacer">
        •
      </span>
      Evangelist
    </span>
  </div>
</div>
`;

exports[`PostListItem React component Renders Normal Post (Blog, Press Release, etc.) correctly 1`] = `
<div
  className="item">
  <div
    className="date">
    <time
      className="time"
      dateTime="2016-12-21T00:00:00-08:00">
      Wed, Dec 21, 2016
    </time>
  </div>
  <a
    className="textlink"
    href={undefined}
    onClick={[Function]}
    style={Object {}}
    target={undefined}>
    <span
      className="title">
      Calculus of the Nervous System
    </span>
  </a>
  <p
    className="paragraph">
    Quick quip or intro to the full text.
    ...
  </p>
  <div
    className="author">
    <span
      className="subtle">
      <span
        className="avatar">
        <img
          alt="Ada Lovelace"
          className="image border round"
          onClick={undefined}
          src="filename.ext"
          title="Ada Lovelace" />
      </span>
      Ada Lovelace
      <span
        className="spacer">
        •
      </span>
      Evangelist
    </span>
  </div>
</div>
`;
